### PR TITLE
plotter: make plotter tests idempotent

### LINKED ISSUE #252: plotter: testing is not idempotent with respect to failure
Invoking `go test` on `github.com/gonum/plotter` is not idempotent if there is a failure since the test overwrites the golden values. This should never happen unless the regen flag is passed in.

The ideal solution is to not write anything to a file system however this is required because of the use of example tests for the general testing. I'm even more convinced now that that requirement is too hard.

### PR DESCRIPTION
**Pull request by kostya-sh**
This change renames all golden files from plotter/testdata to have
"_golden.png" suffix so running tests do not override them.

Fixes #252

### CODE REVIEW

**Comment by ctessum** [comment]
This is more or less what I was planning on doing when I got around to finishing #253. LGTM, thanks.

**Comment by kostya-sh** [comment]
@ctessum. sorry for stealing your PR ;)

**Comment by ctessum** [comment]
No worries! It was making me nervous just sitting there unfinished.

File: plotter/general_test.go [hunk 1]
```diff
@@ -21,30 +21,40 @@ import (
 
 var generateTestData = flag.Bool("regen", false, "Uses the current state to regenerate the test data.")
 
+func goldenPath(path string) string {
+	ext := filepath.Ext(path)
+	noext := strings.TrimSuffix(path, ext)
+	return noext + "_golden" + ext
+}
+
+func readAll(path string) ([]byte, error) {
```
Comment (sbinet) [inline plotter/general_test.go]: AFAICT, this is just `io/ioutil.ReadFile`, isn't it?

File: .gitignore [hunk 2]
```diff
@@ -1,5 +1,7 @@
 **/*.svg
 **/*.png
+!**/testdata/*_golden.png
```
Comment (sbinet) [inline .gitignore]: do we want to ignore `_golden` files ?
I'd say we do want to see via `git status` that a `_golden` file was modified.
and/or see when a new `_golden` file appeared.

File: plotter/general_test.go [hunk 3]
```diff
@@ -53,45 +63,20 @@ func checkPlot(ExampleFunc func(), t *testing.T, paths ...string) {
...
-	for i, path := range paths {
-		errored := false
-		f, err := os.Open(filepath.Join("testdata", path))
-		if err != nil {
-			t.Error(err)
-			errored = true
-		}
-		have, err := ioutil.ReadAll(f)
+	for _, path := range paths {
+		got, err := readAll(path)
```
Comment (sbinet) [inline plotter/general_test.go]: just use `io/ioutil.ReadFile`

File: plotter/general_test.go [hunk 4]
```diff
@@ -53,45 +63,20 @@ func checkPlot(ExampleFunc func(), t *testing.T, paths ...string) {
...
 		if err != nil {
-			t.Error(err)
-			errored = true
+			t.Errorf("Failed to read %s: %s", path, err)
+			continue
 		}
-		err = f.Close()
+		golden := goldenPath(path)
+		want, err := readAll(golden)
```
Comment (sbinet) [inline plotter/general_test.go]: ditto

**Comment by sbinet** [comment]
is this PR completely superseding #253 ?

File: .gitignore (hunk 2 above)
Comment (kostya-sh) [inline .gitignore]: `!` at the beginning of the line tells git to not ignore golden files. Before this change all *.png files were ignored.

File: plotter/general_test.go (hunk 1 above)
Comment (kostya-sh) [inline plotter/general_test.go]: Good point!

**Comment by ctessum** [comment]
Yes.

**Comment by kostya-sh** [comment]
@sbinet , thanks for the review. I've addressed your comments.

File: .gitignore (hunk 2 above)
Comment (sbinet) [inline .gitignore]: ah! sorry, I overlooked that exclamation mark...

**Comment by sbinet** [comment]
LGTM.

**Comment by sbinet** [comment]
@kostya-sh could you send a PR against `gonum/license` so I can merge this PR?
(I don't think you signed it, did you?)

**Comment by kostya-sh** [comment]
@sbinet sent PR against license

**Comment by sbinet** [comment]
needs gonum/license#19

**Comment by sbinet** [comment]
ok, thanks.
I'll wait a day or so for @kortschak to comment (he already LGTM'd #253, so it should be ok.)
@eaburns was also ok with the previous PR...

**Comment by sbinet** [comment]
merged. thanks again.

**Pull request merged**